torch/csrc/jit/codegen/cuda/executor_kernel_arg.cpp

```diff
@@ -0,0 +1,141 @@
+#include <ATen/CUDAGeneratorImpl.h>
+
+// Extract size and strides
+#include <torch/csrc/jit/codegen/cuda/kernel_cache.h>
+
+#include <torch/csrc/jit/codegen/cuda/executor_kernel_arg.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+std::unique_ptr<TensorArgAbstract> getTensorArg(
+    c10::ScalarType dtype,
+    int nDims) {
+  switch (dtype) {
+    case (at::kFloat):
+      return getTensorArg<float>(nDims);
+    case (at::kHalf):
+      return getTensorArg<at::Half>(nDims);
+    case (at::kBool):
+      return getTensorArg<bool>(nDims);
+    default:
+      TORCH_CHECK(
+          false,
+          "Dtype: ",
+          dtype,
+          " not currently supported in code generated kernels.");
+  }
+}
+
+// Push a tensor to the arguments
+void KernelArgumentHolder::push(const at::Tensor& tensor) {
+  changed = true;
+  int nDims = tensor.ndimension();
+
+  c10::ScalarType dtype = tensor.scalar_type();
+  std::unique_ptr<TensorArgAbstract> tensor_arg = getTensorArg(dtype, nDims);
+  tensor_arg->setPointer(tensor.data_ptr());
+  for (int i = 0; i < nDims; i++) {
+    tensor_arg->setSize(i, tensor.sizes()[i]);
+    tensor_arg->setStride(i, tensor.strides()[i]);
+  }
+  arguments.push_back(std::move(tensor_arg));
+}
+
+// Push a tensor to the arguments
+void KernelArgumentHolder::push(
```

Comment:
isn't this with the default null_opt the same as the overload above? Do we need both?